src/cpa/sh.py

```diff
@@ -0,0 +1,384 @@
+            self.read_QM_from_file(istep)
+            if (self.mol.l_qmmm and mm != None):
+                mm.get_data(self.mol, base_dir, bo_list, istep, calc_force_only=False)
+
+            if (not self.mol.l_nacme and self.l_adj_nac):
+                self.mol.adjust_nac()
+
+            self.read_RP_from_file(istep)
+            
+            el_run(self)
+
+            self.hop_prob()
+            self.hop_check(bo_list)
+
+            if (self.dec_correction == "idc"):
+                if (self.l_hop or self.l_reject):
+                    self.correct_dec_idc()
+            elif (self.dec_correction == "edc"):
+                # If kinetic is 0, coefficient/density matrix are update into itself
+                if (self.mol.ekin_qm > eps):
+                    self.correct_dec_edc()
+
+            self.update_energy()
+
+            if ((istep + 1) % self.out_freq == 0):
+                self.write_md_output(unixmd_dir, istep)
+            if ((istep + 1) % self.out_freq == 0 or len(self.event["HOP"]) > 0):
+                self.print_step(istep)
+            
+            self.fstep = istep
+
+            if (not l_save_scr):
+                tmp_dir = os.path.join(unixmd_dir, "scr_qm")
+                if(os.path.exists(tmp_dir)):
+                    shutil.rmtree(tmp_dir)
+
+                if(self.mol.l_qmmm and mm != None):
+                    tmp_dir = os.path.join(unixmd_dir, "scr_mm")
+                    if (os.path.exists(tmp_dir)):
+                        shutil.rmtree(tmp_dir)
+
+    def hop_prob(self):
+        """ Routine to calculate hopping probabilities
+        """
+        # Reset surface hopping variables
+        self.rstate_old = self.rstate
+
+        self.prob = np.zeros(self.mol.nst)
+        self.acc_prob = np.zeros(self.mol.nst + 1)
+
+        self.l_hop = False
+
+        accum = 0.
+
+        for ist in range(self.mol.nst):
+            if (ist != self.rstate):
+                self.prob[ist] = - 2. * self.mol.rho.real[ist, self.rstate] * \
+                    self.mol.nacme[ist, self.rstate] * self.dt / self.mol.rho.real[self.rstate, self.rstate]
+
+                if (self.prob[ist] < 0.):
+                    self.prob[ist] = 0.
+                accum += self.prob[ist]
+            self.acc_prob[ist + 1] = accum
+        psum = self.acc_prob[self.mol.nst]
+
+        if (psum > 1.):
+            self.prob /= psum
+            self.acc_prob /= psum
+
+    def hop_check(self, bo_list):
+        """ Routine to check hopping occurs with random number
+
+            :param integer,list bo_list: List of BO states for BO calculation
+        """
+        self.rand = random.random()
+        for ist in range(self.mol.nst):
+            if (ist == self.rstate):
+                continue
+            if (self.rand > self.acc_prob[ist] and self.rand <= self.acc_prob[ist + 1]):
+                self.l_hop = True
+                self.rstate = ist
+                bo_list[0] = self.rstate
```

Comment:
in line 80, we set `bo_list` to the empty list. so. I think this line is weird...